src/coord/src/coord.rs

```diff
@@ -2040,13 +2085,68 @@ impl Coordinator {
 
     async fn sequence_peek(
         &mut self,
-        conn_id: u32,
+        session: &mut Session,
         source: MirRelationExpr,
         when: PeekWhen,
         finishing: RowSetFinishing,
         copy_to: Option<CopyFormat>,
     ) -> Result<ExecuteResponse, CoordError> {
-        let timestamp = self.determine_timestamp(&source, when)?;
+        let conn_id = session.conn_id();
+        let in_transaction = matches!(
+            session.transaction(),
+            &TransactionStatus::InTransaction(_) | &TransactionStatus::InTransactionImplicit(_)
+        );
+        // For explicit or implicit transactions that do not use AS OF, get the
+        // timestamp of the in-progress transaction or create one. If this is an AS OF
+        // query, we don't care about any possible transaction timestamp. If this is a
+        // single-statement transaction (TransactionStatus::Started), we don't need to
+        // worry about preventing compaction or choosing a valid timestamp for future
+        // queries.
+        let timestamp = if in_transaction && when == PeekWhen::Immediately {
+            let timestamp = session.get_transaction_timestamp(|| {
+                // Determine a timestamp that will be valid for anything in any schema
+                // referenced by the first query. This is a first pass implementation of "time
+                // domains".
+                let ids = self.catalog.timedomain_for(&source, conn_id);
+
+                // We want to prevent compaction of the indexes consulted by
+                // determine_timestamp, not the ones listed in the query.
+                let (timestamp, index_ids) =
+                    self.determine_timestamp(&ids, PeekWhen::Immediately)?;
+                let mut ids = HashSet::new();
+                let mut handles = vec![];
+                for id in index_ids {
+                    handles.push(self.indexes.get(&id).unwrap().since_handle(vec![timestamp]));
+                    ids.insert(id);
+                }
+                self.txn_reads.insert(
+                    conn_id,
+                    TxnReads {
+                        ids,
+                        _handles: handles,
+                    },
+                );
+
+                Ok(timestamp)
+            })?;
+
+            // Verify that the indexes for this query are in the current read transaction.
+            // TODO: we should probably compare the results of timedomain_for instead of nearest_indexes.
+            let (index_ids, _indexes_complete) =
+                self.catalog.nearest_indexes(&source.global_uses());
+            let txn_reads = self.txn_reads.get(&conn_id).unwrap();
+            for id in &index_ids {
+                if !txn_reads.ids.contains(id) {
+                    return Err(CoordError::RelationOutsideTimeDomain(
```

Comment:
This error also happens if a concurrent transaction issues a ```CREATE OR REPLACE VIEW``` for a view that you are selecting from. In this case, the error message is misleading, since the user is operating in a single schema.

```
"materialize.public.v2_primary_idx" not in schemas of the first query
```

If this expected and unavoidable, can we word the error message better, e.g. '"{}" can not be used in this transaction because it did not exist in schema {} or was redefined after the transaction started" or something along those lines (I struggle to come up with a good message in plain English).

Reply:
I'm ok with this error being not exactly specific in the case of a replaced view because otherwise it gets pretty wordy.